cpp/src/lists/copying/gather.cu

```diff
@@ -0,0 +1,192 @@
+  // TODO : it might make sense to expose a gather() function that takes a column_view and
+  // returns a column that does this work correctly.
+  rmm::device_buffer null_mask{0, stream, mr};
+  size_type null_count = column.null_count();
+  if (null_count > 0) {
+    auto list_cdv = column_device_view::create(column);
+    auto validity = cudf::detail::valid_if(
+      gather_map_begin,
+      gather_map_begin + gd.gather_map_size,
+      [cdv = *list_cdv] __device__(int index) { return cdv.is_valid(index) ? true : false; },
+      stream,
+      mr);
+
+    leaf_column->set_null_mask(std::move(validity.first), validity.second);
+  }
+
+  return std::move(leaf_column);
+}
+
+/**
+ * @copydoc cudf::lists::detail::gather_list_nested
+ *
+ */
+std::unique_ptr<column> gather_list_nested(cudf::lists_column_view const& list,
+                                           gather_data& gd,
+                                           cudaStream_t stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  // gather map iterator for this level (N)
+  auto gather_map_begin = thrust::make_transform_iterator(
+    thrust::make_counting_iterator<size_type>(0), list_gatherer{gd});
+  size_type gather_map_size = gd.gather_map_size;
+
+  // gather the bitmask, if relevant
+  rmm::device_buffer null_mask{0, stream, mr};
+  size_type null_count = list.null_count();
+  if (null_count > 0) {
+    auto list_cdv = column_device_view::create(list.parent());
+    auto validity = cudf::detail::valid_if(
+      gather_map_begin,
+      gather_map_begin + gather_map_size,
+      [cdv = *list_cdv] __device__(int index) { return cdv.is_valid(index) ? true : false; },
+      stream,
+      mr);
+    null_mask  = std::move(validity.first);
+    null_count = validity.second;
+  }
+
+  // generate gather_data for next level (N+1), potentially recycling the temporary
+  // base_offsets buffer.
+  gather_data child_gd = make_gather_data<false>(
+    list, gather_map_begin, gather_map_size, stream, mr, std::move(gd.base_offsets));
+
+  // the nesting case.
+  if (list.child().type() == cudf::data_type{type_id::LIST}) {
+    // gather children.
+    auto child = gather_list_nested(list.child(), child_gd, stream, mr);
+
+    // return the nested column
+    return make_lists_column(gather_map_size,
+                             std::move(child_gd.offsets),
+                             std::move(child),
+                             null_count,
+                             std::move(null_mask));
+  }
+
+  // it's a leaf.  do a regular gather
+  auto child = gather_list_leaf(list.child(), child_gd, stream, mr);
+
+  // assemble final column
+  return make_lists_column(gather_map_size,
+                           std::move(child_gd.offsets),
+                           std::move(child),
+                           null_count,
+                           std::move(null_mask));
+}
+
+}  // namespace detail
+}  // namespace lists
+}  // namespace cudf
```

Comment:
Need a blank line here.

Reply:
Arg.  I must be removing these things in my sleep.